src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -1217,4 +1212,37 @@ namespace winrt::TerminalApp::implementation
             itemContainer.DataContext(args.Item());
         }
     }
+
+    // Method Description:
+    // - Reads the list of recent commands from the persistent application state
+    // Return Value:
+    // - The list of FilteredCommand representing the ones stored in the state
+    Windows::Foundation::Collections::IVector<winrt::TerminalApp::FilteredCommand> CommandPalette::_loadRecentCommands() const
+    {
+        const auto parsedCommands = winrt::single_threaded_vector<winrt::TerminalApp::FilteredCommand>();
+        const auto recentCommands = ApplicationState::SharedInstance().RecentCommands();
+        std::for_each(begin(recentCommands), end(recentCommands), [&](const auto& c) {
+            auto parsedCommand = _buildCommandLineCommand(c.c_str());
+            if (parsedCommand.has_value())
+            {
+                parsedCommands.Append(parsedCommand.value());
+            }
+        });
+        return parsedCommands;
+    }
+
+    // Method Description:
+    // - Update recent commands by putting the provided command as most recent.
+    // Upon race condition might override an update made by another window.
+    // Return Value:
+    // - <none>
+    void CommandPalette::_updateRecentCommands(const winrt::hstring& command) const
+    {
+        std::vector<hstring> newRecentCommands;
+        newRecentCommands.push_back(command);
+
+        const auto recentCommands = ApplicationState::SharedInstance().RecentCommands();
+        std::copy_n(begin(recentCommands), std::min(recentCommands.Size(), CommandLineHistoryLength - 1), back_inserter(newRecentCommands));
+        ApplicationState::SharedInstance().RecentCommands(winrt::single_threaded_vector<hstring>(std::move(newRecentCommands)));
```

Comment:
Previously this was accomplished using:
```cpp
if (_commandLineHistory.Size() == CommandLineHistoryLength)
{
    _commandLineHistory.RemoveAtEnd();
}
_commandLineHistory.InsertAt(0, filteredCommand.value());
```

I suspect such an approach to be easier to understand and also to be significantly faster than this unfortunately.

Reply:
It won't be significant faster - because we need still need to reload the commands again (to reduce races between several instances). To improve the readability I will change this to calling the `_loadRecentCommands` and then invoking the previous manipulation.

Reply:
Can you just do the manipulation directly here? e.g.
```
auto recentCommands = ApplicationState::SharedInstance().RecentCommands();
if (recentCommands.Size() >= CommandLineHistoryLength) {
    recentCommands.RemoveAtEnd();
}
recentCommands.InsertAt(0, commands);
```